### DISCUSSION TITLE
Why is turbo sometimes building packages out of order (not using --parallel)?

### DISCUSSION THREAD

**Discussion by jtlapp** [vercel/turborepo#3637] [Help] [answered]
I'm stumped. Turbo seems to be ignoring dependency order, and I'm not using `--parallel`. Sometimes it builds in a workable order, sometimes it doesn't. I'm new to turborepo, attempting to port from an Nx monorepo.

```json
{
  "$schema": "https://turbo.build/schema.json",
  "globalDependencies": ["**/.env.*local"],
  "pipeline": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": ["dist/**", ".next/**"]
    },
    "test": {
      "outputs": ["coverage/**"],
      "dependsOn": []
    },
    "lint": {
      "outputs": []
    },
    "dev": {
      "cache": false
    },
    "clean": {
      "cache": false
    }
  }
}
```

Relevant portion of the dependency graph:

![Screen Shot 2023-02-04 at 08 42 27](https://user-images.githubusercontent.com/138133/216771152-8a6f2b5d-f364-47a3-80a8-526c9660c184.png)

When I run `turbo clean build`:

![Screen Shot 2023-02-04 at 08 43 58](https://user-images.githubusercontent.com/138133/216771168-1f8ed09c-d655-48b2-ad14-3b272e46b188.png)

database/package.json:
```
  "dependencies": {
    "@fieldzoo/config": "*",
    "@fieldzoo/utilities": "*"
  },
```

config/package.json:
```
  "dependencies": {
    "@fieldzoo/utilities": "*",
    "@fieldzoo/validation": "*",
    "@fieldzoo/validators": "*"
  },
```

validation/package.json:
```
  "dependencies": {
    "@fieldzoo/utilities": "*"
  },
```

When it gets the build order right, everything compiles just fine. I can also manually compile them in the proper order without problem.

**Comment by jtlapp** [comment]
I may have solved the problem. Turbo appears to be running the `clean` and `build` commands in parallel in `turbo clean build`. It seems that turbo gets the build order right every time when I do this instead:

```
turbo clean && turbo build
```

(Actually, I have no idea why the above works consistently and the prior does not.)

**Comment by mehulkar** [comment, marked as answer]
`turbo clean build` says "I want to run these tasks, but they aren't related to each other". If you want `clean` to run _before_ `build`, you can add `clean` to the `dependsOn` array of `build` and then just run `turbo run build` (no need to specify `clean` at all from the CLI!)

**Comment by jtlapp** [comment]
Thanks for the clarification. I was only using `clean` after reorganizing the source tree, so I don't need it to precede ever `build`.
